external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -86,27 +112,38 @@ private[kafka010] object CachedKafkaProducer extends Logging {
     paramsSeq
   }
 
-  /** For explicitly closing kafka producer */
+  /** For explicitly closing kafka producer, will not close an inuse producer until released. */
   private[kafka010] def close(kafkaParams: ju.Map[String, Object]): Unit = {
     val paramsSeq = paramsToSeq(kafkaParams)
     guavaCache.invalidate(paramsSeq)
   }
 
-  /** Auto close on cache evict */
-  private def close(paramsSeq: Seq[(String, Object)], producer: Producer): Unit = {
+  /** Close this producer and process pending closes. */
+  private def close(producer: CachedKafkaProducer): Unit = synchronized {
     try {
-      logInfo(s"Closing the KafkaProducer with params: ${paramsSeq.mkString("\n")}.")
       producer.close()
+      // Check and close any producers evicted, and pending to be closed.
+      for (p <- closeQueue.iterator().asScala) {
+        if (p.inUseCount.intValue() <= 0) {
+          producer.close()
```

Comment:
Could you print the producer UUID here as well for tracking reasons?

Reply:
I want to be consistent too, if everyone agrees one way, I can change it. Currently, I have favoured less verbosity.

Reply:
Sorry, wrote the message at the wrong place. You can ignore it.